### ISSUE TITLE
Question about FedAvg strategy.

### ISSUE THREAD

**Issue by Arctic-Xiangjian** [owkin/FLamby#260]
Hello, I find that the fedavg code only consider the molde.parameters() to average. And due to the BN layer can not be averaged the FedAvg will actuallly give n different models (n is the number of clients).  It's actually like this paper's approch which they call as Siloed Federated Learning. (https://arxiv.org/pdf/2008.07424.pdf)

I test my code to average all layers, it seems will give higher performance of IXI dataset and also gives a more fair baseline.

```
# Aggregation step
weight_keys=self.models_list[0].model.state_dict().keys()
for key in weight_keys:
    key_avg=0
    for _model,train_size in zip(self.models_list,self.training_sizes):
        key_avg=key_avg+_model.model.state_dict()[key]*train_size/float(self.total_number_of_samples)
    fed_state_dict[key]=key_avg
for _model in self.models_list:
    _model.model.load_state_dict(fed_state_dict)
```

**Comment by jeandut** [comment]
Hello @Arctic-Xiangjian, you are completely right we do not average BN parameters thus effectively creating $n$ different models in the sense that they all share the same parameters but have different BN statistics.
However for evaluation we always use the first model so in a way we shoot ourselves in the foot by doing that so our baseline is "fair". 
In fact there is no general consensus in the literature on what to do with BN layers in FL settings, you have FedBN (https://arxiv.org/abs/2102.07623) or the SiloedBN work you mentioned (disclaimer I am first author of this paper) and some other approaches but usually in the literature most people use non-batch-wise normalization techniques.
However with siloedBN we have a more private instantiation of Federated Averaging because BN statistics are key to state-of-the-art data reconstruction attacks.
This repository is not meant to force everyone to use one technique or the other but to give the user the liberty to choose what she or he think is more relevant.
Maybe it would be nice to pass how to average BN statistics as an option to the strategy ?

**Comment by Arctic-Xiangjian** [comment]
> Hello @Arctic-Xiangjian, you are completely right we do not average BN parameters thus effectively creating n different models in the sense that they all share the same parameters but have different BN statistics. However for evaluation we always use the first model so in a way we shoot ourselves in the foot by doing that so our baseline is "fair". In fact there is no general consensus in the literature on what to do with BN layers in FL settings, you have FedBN (https://arxiv.org/abs/2102.07623) or the SiloedBN work you mentioned (disclaimer I am first author of this paper) and some other approaches but usually in the literature most people use non-batch-wise normalization techniques. However with siloedBN we have a more private instantiation of Federated Averaging because BN statistics are key to state-of-the-art data reconstruction attacks (https://arxiv.org/abs/2104.07586). This repository is not meant to force everyone to use one technique or the other but to give the user the liberty to choose what she or he think is more relevant. Maybe it would be nice to be able to pass how to average BN statistics as an option to the strategy ?

Thank you for detailed reply~ The reconstruction paper is very helpful, I am very interested in this direction. 

And I was wondering if we do not average the BN layers, we would have different models. May be it will be better when we test the result on each dataset, we can use their own model insteated of Model 0?(If we have a big gap thought different clinents this might be more fair.)

**Comment by jeandut** [comment]
Your idea to use each model on its own distribution is exactly what we are doing in SiloedBN and very close to works on personalization in FL.
It would indeed be probably better in terms of performances but is not what we implemented in the article. Feel free to test it by running something along the lines of (similarly as [the personalization example](https://github.com/owkin/FLamby/blob/main/flamby/personalization_example/compare_fedavg_vs_perso.py) you can find in the repository):
```python
models = strat(**kwargs).run()
perfs = []
for i in range(n_clients):
    perf_dict = evaluate_model_on_tests(models[i], [test_dls[i]], metric)
    perfs.append(perf_dict["client_0"])
```

**Comment by jeandut** [comment]
@Arctic-Xiangjian closing now that the question is answered. Feel free to reopen an issue if needed.
